source/components/composing-components.md

```diff
@@ -0,0 +1,260 @@
+Sending data down allows the consumer of our component to utilize that data.
+For the consumer to get access to the data, she needs to know what data a component exposes,
+this is where documentation is so crucial, and from there the yielded data can be accessed
+with the `as` operator. Let's take the following template as an example:
+
+```app/components/user-list/template.hbs
+{{#each users as |user|}}
+  {{yield user}}
+{{/each}}
+```
+
+This `{{user-list}}` component will yield as many times as there are users.
+We can consume it in the following way:
+
+```app/users/template.hbs
+{{#user-list users=users as |user|}}
+  {{user-card user=user}}
+{{/user-list}}
+```
+
+Now `{{user-card}}` has access to the current user, which changes as the `{{user-list}}` iterates over each user
+since the `{{yield user}}` is inside the `{{each}}` block. Although this is nice, what if we used our component
+without a block, i.e. `{{user-list users=users}}`? This would make the component pretty useless since nothing is yielding,
+but the users are still being iterated. Let's mix in the `hasBlock` attribute and see if we can make it more useful.
+
+```app/components/user-list/template.hbs
+{{#if hasBlock}}
+  {{#each users as |user|}}
+    {{yield user}}
+  {{/each}}
+{{else}}
+  {{#each users as |user|}}
+    <section class="user-info">
+      <h3>{{user.name}}</h3>
+      <p>{{user.bio}}</p>
+    </section>
+  {{/each}}
+{{/if}}
+```
+
+Now we our component is a little more useful, because it has sane defaults, but it also allows us to override
+those defaults. I'm sure there is more that we can do to our `{{user-list}}` component to make it more useful,
+like handling different types of users.
+
+#### `{{component}}` Helper
+
+Now we need to add a new type of user, a superuser, which will have different behavior in our
+application. How would the consumer use this new data to show a different UI for each type of user?
+This is where the `{{component}}` helper comes into play. This helper allows us to render a component
+based on an arbitrary value. Along with this new feature request, we were also asked to add a toggle
+between a basic and a detailed view, which we need to expose in our block.
+
+Since components require a dash in the component name, e.g. `basic-card-superuser`,
+the simplest way would be to use a helper to concatenate 'basic-card-' and `user.type`.
+
+```app/users/template.hbs
+{{#user-list users=users as |user basicMode|}}
+  {{#if basicMode}}
+    {{component (concat 'basic-card-' user.type) user=user}}
+  {{else}}
+    {{component (concat 'detailed-card-' user.type) user=user}}
+  {{/if}}
+{{/user-list}}
+```
+
+Now we have `{{basic-card-superuser}}` and `{{detailed-card-public}}` rendering in their respective places
+based on the `user.type` equaling 'superuser' and 'public' respectively. Here we used the 'concat' helper
+as a subexpression inside of our `{{component}}` helper. Subexpressions are evaluated before the helper
```

Comment:
Please use the phrase `nested helper` instead of `subexpression`. We believe `nested helper` may be more friendly to those not brainwashed by the htmlbars hype ;-)